core/src/main/java/com/datastax/oss/driver/internal/core/metrics/DefaultNodeMetricUpdater.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.metrics;
+
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.MetricRegistry;
+import com.datastax.oss.driver.api.core.metadata.Node;
+import com.datastax.oss.driver.api.core.metrics.CoreNodeMetric;
+import com.datastax.oss.driver.api.core.metrics.NodeMetric;
+import com.datastax.oss.driver.internal.core.context.InternalDriverContext;
+import com.datastax.oss.driver.internal.core.pool.ChannelPool;
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.util.Set;
+
+public class DefaultNodeMetricUpdater implements NodeMetricUpdater {
+
+  private final String metricNamePrefix;
+  private final MetricRegistry metricRegistry;
+  private final Set<NodeMetric> enabledMetrics;
+
+  public DefaultNodeMetricUpdater(
+      Node node, Set<NodeMetric> enabledMetrics, InternalDriverContext context) {
+    this.metricNamePrefix = buildPrefix(context.sessionName(), node.getConnectAddress());
+    this.enabledMetrics = enabledMetrics;
+    this.metricRegistry = context.metricRegistry();
+
+    if (enabledMetrics.contains(CoreNodeMetric.pooled_connection_count)) {
+      metricRegistry.register(
+          metricNamePrefix + CoreNodeMetric.pooled_connection_count.name(),
+          (Gauge<Integer>)
+              () -> {
+                ChannelPool pool = context.poolManager().getPools().get(node);
+                return (pool == null) ? 0 : pool.size();
+              });
+    }
+    if (enabledMetrics.contains(CoreNodeMetric.available_stream_count)) {
+      metricRegistry.register(
+          metricNamePrefix + CoreNodeMetric.available_stream_count,
+          (Gauge<Integer>)
+              () -> {
+                ChannelPool pool = context.poolManager().getPools().get(node);
+                return (pool == null) ? 0 : pool.getAvailableIds();
+              });
+    }
+  }
+
+  @Override
+  public void markMeter(NodeMetric metric) {
+    if (enabledMetrics.contains(metric)) {
+      metricRegistry.meter(metricNamePrefix + metric.name()).mark();
```

Comment:
Yes, there will be a method for each metric type: `markMeter`,  `updateTimer`, `incrementCounter`, etc.
The caller is expected to know what type of metric each enum constant corresponds to. A mistake is very unlikely because it's all driver code, and the same person will add a metric and add the code to update it. And if a mistake does happen the error should be pretty obvious.

The problem with something like `NodeMetric#update` is that `NodeMetric` couldn't be an enum anymore (because it would need a reference to the `Meter`, which is created at runtime, and also because there could be multiple instances for separate sessions). Then the list of metric names wouldn't be known statically, we'd probably need a `NodeMetricRegistry` in order to validate the configuration, etc. The (very low) risk of a mistake does not justify all that complexity IMO.

Reply:
Sounds reasonable. 👍